domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -48,13 +57,24 @@ class ExplorationProgressController @Inject constructor(
   //  performing a post-load operation. The current state of the controller is leaking this
   //  implementation detail to tests.
 
+  /** Indicates that the checkpoint database has exceeded the allocated limit. */
+  class CheckpointDatabaseExceededLimitException(msg: String) : Exception(msg)
+
+  /** Indicates that exploration contains unsaved progress. */
+  class ProgressNotSavedException(msg: String) : Exception(msg)
+
   private val currentStateDataProvider =
     dataProviders.createInMemoryDataProviderAsync(
       CURRENT_STATE_DATA_PROVIDER_ID,
       this::retrieveCurrentStateAsync
     )
   private val explorationProgress = ExplorationProgress()
   private val explorationProgressLock = ReentrantLock()
+  private val explorationCheckpointDataProvider =
+    dataProviders.createInMemoryDataProviderAsync(
+      EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID,
+      this::saveCheckpointAsync
+    )
```

Comment:
If I observe the `DataProvider` in StateFragmentPresenter the checkpoint is getting saved, if I remove the observer checkpoints are not getting saved.

Do `DataProviders` don't perform any operation until they are observed? 
If so wouldn't it be better to communicate back to the domain layer from here instead of creating a new `DataProvider` and using `.retrieveData`